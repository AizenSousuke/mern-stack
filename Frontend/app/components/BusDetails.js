import moment from "moment";
import React, { useState, useEffect } from "react";
import { View, Text } from "react-native";
import * as Progress from "react-native-progress";
import styles from "../../assets/css/AppStyles";

const GetValueForLoad = (load) => {
	switch (load) {
		case "SEA":
			return 0.2;
		case "SDA":
			return 0.6;
		default:
			return 0.9;
	}
};

const GetColorForLoad = (load) => {
	switch (load) {
		case "SEA":
			return "green";
		case "SDA":
			return "orange";
		default:
			return "red";
	}
};

const GetEstimatedArrivalMinute = (estimatedArrivalTime) => {
	var timeNow = moment().utcOffset("+08:00").format();
	var time1 = moment(timeNow);
	var time2 = moment(estimatedArrivalTime);
	// console.log("Time now: " + timeNow);
	// console.log("Time bus is reaching: " + estimatedArrivalTime);
	var timeEstimatedArrival = time2.diff(time1, "minutes");
	// console.log("Time estimated arrival: " + timeEstimatedArrival);

	if (parseInt(timeEstimatedArrival) > 2) {
		return timeEstimatedArrival + " min";
	} else {
		return "Arrived";
	}
};

const BusDetails = ({ busNumber, details }) => {
	// console.log("Details: " + JSON.stringify(details));

	const [number, setNumber] = useState(busNumber);
	const [data, setData] = useState(details);

	useEffect(() => {
		setNumber(busNumber);
		setData(details);
	}, [busNumber, details]);

	return (
		<View style={styles.busDetails}>
			<Text style={[styles.busNumber, { flex: 0.2 }]}>{number}</Text>
			<View
				style={{
					flex: 0.8,
					flexDirection: "row",
					justifyContent: "space-around",
				}}
			>
				<View style={{ flexDirection: "column" }}>
					{data.NextBus.Load !== "" ? (
						<>
							<Text style={styles.busType}>
								{data.NextBus.Type == "DD" ? "Double" : ""}
							</Text>
							<Progress.Bar
								progress={GetValueForLoad(data.NextBus.Load)}
								color={GetColorForLoad(data.NextBus.Load)}
								width={50}
							/>
							<Text style={styles.estimatedArrival}>
								{GetEstimatedArrivalMinute(
									data.NextBus.EstimatedArrival
								).toString()}
							</Text>
						</>
					) : (
						<Text style={styles.noData}>No Data</Text>
					)}
				</View>
				<View style={{ flexDirection: "column" }}>
					{data.NextBus2.Load !== "" ? (
						<>
							<Text style={styles.busType}>
								{data.NextBus2.Type == "DD" ? "Double" : ""}
							</Text>
							<Progress.Bar
								progress={GetValueForLoad(
									data.NextBus2.Load
								)}
								color={GetColorForLoad(data.NextBus2.Load)}
								width={50}
							/>
							<Text style={styles.estimatedArrival}>
								{GetEstimatedArrivalMinute(
									data.NextBus2.EstimatedArrival
								).toString()}
							</Text>
						</>
					) : (
						<Text style={styles.noData}>No Data</Text>
					)}
				</View>
				<View style={{ flexDirection: "column" }}>
					{data.NextBus3.Load !== "" ? (
						<>
							<Text style={styles.busType}>
								{data.NextBus2.Type == "DD" ? "Double" : ""}
							</Text>
							<Progress.Bar
								progress={GetValueForLoad(
									data.NextBus3.Load
								)}
								color={GetColorForLoad(data.NextBus3.Load)}
								width={50}
							/>
							<Text style={styles.estimatedArrival}>
								{GetEstimatedArrivalMinute(
									data.NextBus3.EstimatedArrival
								).toString()}
							</Text>
						</>
					) : (
						<Text style={styles.noData}>No Data</Text>
					)}
				</View>
			</View>
		</View>
	);
};

export default BusDetails;
