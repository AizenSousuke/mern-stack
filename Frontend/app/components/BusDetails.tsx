import moment from "moment";
import React, { useState, useEffect } from "react";
import { View, Text } from "react-native";
import * as Progress from "react-native-progress";
import AppStyles from "../../assets/css/AppStyles";
import BusLocation from "./BusLocation";
import BusInformation from "./BusInformation";
import ModalAdder from "./ModalAdder";

const GetValueForLoad = (load: string) => {
	switch (load) {
		case "SEA":
			return 0.2;
		case "SDA":
			return 0.6;
		default:
			return 0.9;
	}
};

const GetColorForLoad = (load: string) => {
	switch (load) {
		case "SEA":
			return "green";
		case "SDA":
			return "orange";
		default:
			return "red";
	}
};

const GetEstimatedArrivalMinute = (estimatedArrivalTime: string) => {
	var timeNow = moment().utcOffset("+08:00").format();
	var time1 = moment(timeNow);
	var time2 = moment(estimatedArrivalTime);
	// console.log("Time now: " + timeNow);
	// console.log("Time bus is reaching: " + estimatedArrivalTime);
	var timeEstimatedArrival: number = time2.diff(time1, "minutes");
	// console.log("Time estimated arrival: " + timeEstimatedArrival);

	if (parseInt(timeEstimatedArrival.toString()) > 2) {
		return timeEstimatedArrival + " min";
	} else {
		return "Arrived";
	}
};

const BusDetails = ({
	busStopData,
	busNumber,
	details,
	busStopLocation,
}: {
	busStopData: any;
	busNumber: string;
	details: any;
	busStopLocation: any;
}) => {
	const [number, setNumber] = useState(busNumber);
	const [data, setData] = useState(details);

	useEffect(() => {
		setNumber(busNumber);
		setData(details);
	}, [busNumber, details]);

	return (
		<View style={AppStyles.busDetails}>
			<ModalAdder
				modalElement={
					<BusInformation
						busNumber={number}
						busStopCode={busStopData?.BusStopCode}
					/>
				}
				style={{
					flex: 0.2,
				}}
			>
				<View style={{ flexDirection: "column" }}>
					<Text style={[AppStyles.busNumber]}>{number}</Text>
				</View>
			</ModalAdder>
			<ModalAdder
				modalElement={
					<BusLocation
						busNumber={details?.ServiceNo}
						nextBus={details?.NextBus}
						busStopLocation={busStopLocation}
					/>
				}
				style={AppStyles.modal}
			>
				<View style={{ flexDirection: "column" }}>
					{data?.NextBus?.Load ? (
						<>
							<Text style={AppStyles.busType}>
								{data?.NextBus?.Type == "DD" ? "Double" : ""}
							</Text>
							<Progress.Bar
								progress={GetValueForLoad(data?.NextBus?.Load)}
								color={GetColorForLoad(data?.NextBus?.Load)}
								width={50}
							/>
							<Text style={AppStyles.estimatedArrival}>
								{GetEstimatedArrivalMinute(
									data?.NextBus?.EstimatedArrival
								).toString()}
							</Text>
						</>
					) : (
						<Text style={AppStyles.noData}>No Data</Text>
					)}
				</View>
				<View style={{ flexDirection: "column" }}>
					{data?.NextBus2?.Load ? (
						<>
							<Text style={AppStyles.busType}>
								{data?.NextBus2?.Type == "DD" ? "Double" : ""}
							</Text>
							<Progress.Bar
								progress={GetValueForLoad(data?.NextBus2?.Load)}
								color={GetColorForLoad(data?.NextBus2?.Load)}
								width={50}
							/>
							<Text style={AppStyles.estimatedArrival}>
								{GetEstimatedArrivalMinute(
									data?.NextBus2?.EstimatedArrival
								).toString()}
							</Text>
						</>
					) : (
						<Text style={AppStyles.noData}>No Data</Text>
					)}
				</View>
				<View style={{ flexDirection: "column" }}>
					{data?.NextBus3?.Load ? (
						<>
							<Text style={AppStyles.busType}>
								{data?.NextBus2?.Type == "DD" ? "Double" : ""}
							</Text>
							<Progress.Bar
								progress={GetValueForLoad(data?.NextBus3?.Load)}
								color={GetColorForLoad(data?.NextBus3?.Load)}
								width={50}
							/>
							<Text style={AppStyles.estimatedArrival}>
								{GetEstimatedArrivalMinute(
									data?.NextBus3?.EstimatedArrival
								).toString()}
							</Text>
						</>
					) : (
						<Text style={AppStyles.noData}>No Data</Text>
					)}
				</View>
			</ModalAdder>
		</View>
	);
};

export default BusDetails;
