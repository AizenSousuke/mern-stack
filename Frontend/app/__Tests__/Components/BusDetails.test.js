import BusDetails from "../../components/BusDetails";
import { render } from "@testing-library/react-native";

describe("Bus Details", () => {
	const details = {
		ServiceNo: 911,
	};
	it("renders correctly", () => {
		render(
			<BusDetails
				busNumber={"911"}
				busStopData={null}
				busStopLocation={null}
				details={details}
			/>
		);
	});
});
