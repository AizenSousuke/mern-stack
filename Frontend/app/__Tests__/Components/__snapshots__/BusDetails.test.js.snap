// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bus Details renders BusDetails according to snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "white",
      "flexDirection": "row",
      "justifyContent": "space-around",
      "padding": 5,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "flex": 0.2,
      }
    }
    testID="Pressable"
  >
    <Modal
      hardwareAccelerated={false}
      onRequestClose={[Function]}
      theme={
        {
          "colors": {
            "black": "#242424",
            "disabled": "hsl(208, 8%, 90%)",
            "divider": "#bcbbc1",
            "error": "#ff190c",
            "grey0": "#393e42",
            "grey1": "#43484d",
            "grey2": "#5e6977",
            "grey3": "#86939e",
            "grey4": "#bdc6cf",
            "grey5": "#e1e8ee",
            "greyOutline": "#bbb",
            "platform": {
              "android": {
                "error": "#f44336",
                "grey": "rgba(0, 0, 0, 0.54)",
                "primary": "#2196f3",
                "searchBg": "#dcdce1",
                "secondary": "#9C27B0",
                "success": "#4caf50",
                "warning": "#ffeb3b",
              },
              "default": {
                "error": "#ff3b30",
                "grey": "#7d7d7d",
                "primary": "#007aff",
                "searchBg": "#dcdce1",
                "secondary": "#5856d6",
                "success": "#4cd964",
                "warning": "#ffcc00",
              },
              "ios": {
                "error": "#ff3b30",
                "grey": "#7d7d7d",
                "primary": "#007aff",
                "searchBg": "#dcdce1",
                "secondary": "#5856d6",
                "success": "#4cd964",
                "warning": "#ffcc00",
              },
              "web": {
                "error": "#ff190c",
                "grey": "#393e42",
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#ca71eb",
                "success": "#52c41a",
                "warning": "#faad14",
              },
            },
            "primary": "#2089dc",
            "searchBg": "#303337",
            "secondary": "#ca71eb",
            "success": "#52c41a",
            "warning": "#faad14",
            "white": "#ffffff",
          },
        }
      }
      transparent={true}
      visible={false}
    />
    <View
      style={
        {
          "flexDirection": "column",
        }
      }
    >
      <Text
        style={
          [
            {
              "alignSelf": "center",
              "color": "#370617",
              "fontSize": 20,
              "fontWeight": "bold",
              "textAlign": "center",
            },
          ]
        }
      >
        911
      </Text>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "flex": 0.8,
        "flexDirection": "row",
        "justifyContent": "space-evenly",
      }
    }
    testID="Pressable"
  >
    <Modal
      hardwareAccelerated={false}
      onRequestClose={[Function]}
      theme={
        {
          "colors": {
            "black": "#242424",
            "disabled": "hsl(208, 8%, 90%)",
            "divider": "#bcbbc1",
            "error": "#ff190c",
            "grey0": "#393e42",
            "grey1": "#43484d",
            "grey2": "#5e6977",
            "grey3": "#86939e",
            "grey4": "#bdc6cf",
            "grey5": "#e1e8ee",
            "greyOutline": "#bbb",
            "platform": {
              "android": {
                "error": "#f44336",
                "grey": "rgba(0, 0, 0, 0.54)",
                "primary": "#2196f3",
                "searchBg": "#dcdce1",
                "secondary": "#9C27B0",
                "success": "#4caf50",
                "warning": "#ffeb3b",
              },
              "default": {
                "error": "#ff3b30",
                "grey": "#7d7d7d",
                "primary": "#007aff",
                "searchBg": "#dcdce1",
                "secondary": "#5856d6",
                "success": "#4cd964",
                "warning": "#ffcc00",
              },
              "ios": {
                "error": "#ff3b30",
                "grey": "#7d7d7d",
                "primary": "#007aff",
                "searchBg": "#dcdce1",
                "secondary": "#5856d6",
                "success": "#4cd964",
                "warning": "#ffcc00",
              },
              "web": {
                "error": "#ff190c",
                "grey": "#393e42",
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#ca71eb",
                "success": "#52c41a",
                "warning": "#faad14",
              },
            },
            "primary": "#2089dc",
            "searchBg": "#303337",
            "secondary": "#ca71eb",
            "success": "#52c41a",
            "warning": "#faad14",
            "white": "#ffffff",
          },
        }
      }
      transparent={true}
      visible={false}
    />
    <View
      style={
        {
          "flexDirection": "column",
        }
      }
    >
      <Text
        style={
          {
            "paddingTop": 8,
          }
        }
      >
        No Data
      </Text>
    </View>
    <View
      style={
        {
          "flexDirection": "column",
        }
      }
    >
      <Text
        style={
          {
            "paddingTop": 8,
          }
        }
      >
        No Data
      </Text>
    </View>
    <View
      style={
        {
          "flexDirection": "column",
        }
      }
    >
      <Text
        style={
          {
            "paddingTop": 8,
          }
        }
      >
        No Data
      </Text>
    </View>
  </View>
</View>
`;
